ENH add `from_cv_results` in `RocCurveDisplay` (single `RocCurveDisplay`)

lucyleeow opened the pull request:
#### Reference Issues/PRs

#### Reference Issues/PRs
Supercedes #25939

This is part of a group of draft PRs to determine best API for adding plots for cv results to our displays.

* Add multi display class (#30359) 
* Use list of single display classes (#30370)
* Amend single display class to optionally return list (this PR)

For all 3 options we take the output of `cross_validate`, and use the fitted estimator and test indicies. No fitting is done in the display.

We **do** recalculate the predictions (which would have already been done in `cross_validate`), which could be avoided if we decided to change `cross_validate` to optionally return the predictions as well (note this would make `cross_val_predict` redundant).
See more thread: https://github.com/scikit-learn/scikit-learn/pull/25939#discussion_r1149423175). I think should be outside of the scope of this body of work though.

#### What does this implement/fix? Explain your changes.
Not 100% I've implemented this optimally.
* `RocCurveDisplay` object may contain data (fpr/tpf etc) for single or multi curves
* `RocCurveDisplay` returns single mpl Artist object, or list of objects for multi curves
* `RocCurveDisplay.plot` handles both single and multi-curve plotting, this has meant a lot more checking is required (c.f. the other 2 implementations, as this is the only case where you can use `plot` directly to plot multi-curves)

More specific concerns detailed in review comments

Plot looks like:

![image](https://github.com/user-attachments/assets/74e2295b-d64e-4d40-afd7-f9341fd90578)



## TODO

We should update `visualization.rst` after this PR is in to add a section about `from_cv_results`.

lucyleeow committed:
first commit

lucyleeow committed:
lint

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -449,3 +525,194 @@ def from_predictions(
+            )
+            roc_auc = auc(fpr, tpr)
+            # Append all
+            fpr_all.append(fpr)
+            tpr_all.append(tpr)
+            auc_all.append(roc_auc)
```
There may be a better way to do this?

jeremiedbb replied:
you could do something like 
```py
fpr_tpr_auc = []

for ...:
    fpr_tpr_auc.append((fpr, tpr, auc))

fpr_all, tpr_all_ auc_all = zip(*fpr_tpr_auc)
```
but I don't know it's really a "better" way :)

lucyleeow replied:
There are fewer lines, so nicer in one aspect at least!

lucyleeow commented on sklearn/utils/_plotting.py:
```diff
@@ -17,14 +17,20 @@ class _BinaryClassifierCurveDisplayMixin:
 
-        name = self.estimator_name if name is None else name
+        # Not 100% sure on this change
+        if n_multi is None:
+            name = self.estimator_name if name is None else name
+        else:
+            name = [f"{curve_type} fold {curve_idx}:" for curve_idx in range(n_multi)]
```
If we go with this implementation, I thought this change could be used for other multi cv displays. Not 100% sure on this change though.

jeremiedbb commented:
As discussed in today's meeting, this is my favorite solution because it's the simplest and least surprising one from a user point of view, even though it adds a bit more internal complexity than the others. And I think we can mitigate some of it by extracting parts of the `plot` code into dedicated `_plot_single` and `_plot_multiple` methods. Or just into small helpers, that would already help readability.

It also looks like a good portion of the added complexity will be exactly the same for other displays like PRCurveDisplay, so there might be a chance that we'll be able to factorize some parts to be used by several displays.

lucyleeow committed:
fixes

lucyleeow committed:
factorize

lucyleeow committed:
fix

lucyleeow commented:
The changes in [f0908e1](https://github.com/scikit-learn/scikit-learn/pull/30399/commits/f0908e1e7d14d5bf1c557b3b438c7302bed7ca9e) and [7e77d4c](https://github.com/scikit-learn/scikit-learn/pull/30399/commits/7e77d4c72307e6ea709624606ad84867af0a4652)  factorizes out common code (compared to #30508), adding helper function to either `_BinaryClassifierCurveDisplayMixin` (if function relevant to other binary displays) or `sklearn/utils/_plotting.py` (if function more generally applicable to more diplays - these potentially could be a parent class method?)

lucyleeow committed:
review changes

lucyleeow committed:
merge main

lucyleeow committed:
lint

lucyleeow committed:
ignore mypy

lucyleeow committed:
fix validate plot param

lucyleeow committed:
fix from predictions

lucyleeow committed:
fix example in docstring

lucyleeow committed:
fix tests

lucyleeow committed:
fix docstring example

lucyleeow committed:
black

lucyleeow committed:
black

lucyleeow committed:
fix tests

lucyleeow committed:
fix testst

lucyleeow committed:
fix tests

lucyleeow committed:
lint

lucyleeow commented:
Note to self, a good example to update to using this new enh would be:  #29611

lucyleeow committed:
fix docstring

lucyleeow committed:
Merge branch 'main' into cv_results3

lucyleeow committed:
add whats new

lucyleeow committed:
amend current test

lucyleeow committed:
lint

lucyleeow committed:
lint

lucyleeow commented on sklearn/metrics/_plot/tests/test_common_curve_display.py:
```diff
@@ -176,9 +174,9 @@ def test_display_curve_estimator_name_multiple_calls(
-)
+# Add separate test for displays that have converted to names?,
+# add note to remove this one in 1.9
+@pytest.mark.parametrize("Display", [DetCurveDisplay, PrecisionRecallDisplay])
```
I was thinking that I could add a separate test that uses `names` (instead of `estimator_name`), so we can delete this test once we've switched all these estimators across to `names`?
This avoids complicate if/else inside this test.
WDYT?

lucyleeow commented on sklearn/metrics/_plot/tests/test_roc_curve_display.py:
```diff
@@ -302,7 +304,7 @@ def test_plot_roc_curve_pos_label(pyplot, response_method, constructor_name):
     classifier = LogisticRegression()
```
I thought about adding `from_cv_results` to this (`test_plot_roc_curve_pos_label`) test BUT I see we create an imbalanced training set, and pass a separate test set to `from_predictions` and `from_estimator`, which we could not do with `from_cv_results`, so it feels like it would miss the spirit of this test. I think we should add a separate test for `pos_label` `from_cv_results`, WDYT?

lucyleeow commented on sklearn/metrics/_plot/tests/test_roc_curve_display.py:
```diff
@@ -302,7 +304,7 @@ def test_plot_roc_curve_pos_label(pyplot, response_method, constructor_name):
     classifier.fit(X_train, y_train)
 
-    # sanity check to be sure the positive class is classes_[0] and that we
+    # sanity check to be sure the positive class is `classes_[0]` and that we
     # are betrayed by the class imbalance
```
@glemaitre I don't quite follow this comment?

lucyleeow replied:
But what does "are betrayed by the class imbalance" mean? How does class imbalance affect `pos_label` ?

lucyleeow replied:
I've amended the note, but let me know if it makes sense!

DeaMariaLeon commented:
As a (not so educated) user, I wonder if it wouldn't be better to add different colours to each plot line? (they all look blue now). I see that here it was decided to do that here: https://github.com/scikit-learn/scikit-learn/pull/30508#discussion_r1914297281.

DeaMariaLeon commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -41,10 +55,44 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
 
         .. versionadded:: 0.24
 
+    fpr : ndarray or list of ndarray
```
Does `fpr` also take a list of ndarray? I think that it gets converted to a list on `_deprecate_singular`. As a "user" I would be a bit confused there I think.

lucyleeow replied:
That was a copy pasta! Fixed now.

These are deprecated now - they'll have the red deprecated box (via the `.. deprecated`), which I am hoping will be enough.

If the user does use it, they will get a warning and what they input will be put into a list and passed. This is all stated in the warning. I thought about adding it to the docstring here as well, but thought it wasn't necessary and would make the parameter descriptions un-necessarily long and complicated. Open to other opinions though!

DeaMariaLeon commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -41,10 +55,44 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
+            `fpr` is deprecated in 1.7 and will be removed in 1.9.
+            Use `fprs` instead.
+
+    tpr : ndarray or list of ndarray
```
same here, does it take a list too?

DeaMariaLeon commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -75,43 +130,61 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
     >>> pred = np.array([0.1, 0.4, 0.35, 0.8])
     >>> fpr, tpr, thresholds = metrics.roc_curve(y, pred)
```
Do `fpr`and `tpr` change to plural (`fprs`and `tprs`) only after they are officially deprecated?

lucyleeow replied:
The singular parameters will be deprecated in 1.7, and then completely removed in 1.9. Between this time users will still be able to use the old singular parameter but they will be given a deprecation warning.

Bit hard to find, but some more details about deprecation here: https://scikit-learn.org/dev/developers/contributing.html#deprecation

DeaMariaLeon replied:
Thank you for the link to deprecation. 
What I wanted to say was that a new user will see that `fpr` and `tpr` are deprecated (warnings and descriptions in the docs).. But then see them being used or included in the example. 
The deprecation link doesn't specify what to use in the examples I think.
 
It's truly a "microscopic" detail. We can discuss it during the call if you think we should. Otherwise you can just ignore me. 🙂

jeremiedbb replied:
> What I wanted to say was that a new user will see that fpr and tpr are deprecated (warnings and descriptions in the docs).. But then see them being used or included in the example.

We actually check that the examples don't raise future warnings so we need to replace them by the new names in the doc. Pushing a commit with `[doc build]` will run all the examples and show all occurrences to update. But it's not urgent, it can be done in a later stage of the PR when all the names and API details are final.

lucyleeow replied:
Oh I see what you are referring to now.

I do use the new `RocCurveDisplay` parameters (the doc testing did run the docstring examples):

```
    >>> display = metrics.RocCurveDisplay(fprs=[fpr], tprs=[tpr], roc_aucs=[roc_auc],
    ...                                   names=['example estimator'])
```

but in the above line:

```
fpr, tpr, thresholds = metrics.roc_curve(y, pred)
```
we do only get one set of `fpr` and `tpr` out so I thought I would leave these are singular. Open to change though.

lucyleeow committed:
review

lucyleeow commented:
I'm happy to keep singular name, prevents deprecation!

You would allow both single ndarray and list of ndarray input right?

lucyleeow committed:
review

lucyleeow commented:
(FYI I'll wait for the rest of your comments and make changes, and then add tests, thanks!)

glemaitre commented:
I'll give another review on this one. @jeremiedbb do you have additional feedback for @lucyleeow?

lucyleeow committed:
Merge branch 'main' into cv_results3

lucyleeow committed:
revert to singular only

lucyleeow committed:
lint

lucyleeow committed:
update tests, more fixes

lucyleeow commented:
I've amended the PR to keep all singular parameter names, which can take a single ndarrary/str etc or a list of such.

The attributes are processed such that they are always a list or `None` and named with trailing `_` e.g., `self.tpr_`

This has avoided a lot of deprecation, which is nice!

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -368,7 +423,7 @@ def from_predictions(
         name : str, default=None
-            Name of ROC curve for labeling. If `None`, name will be set to
+            Name of ROC curve for legend labeling. If `None`, name will be set to
             `"Classifier"`.
```
I am wondering it this is still a good default, to name things "Classifier". Not sure if it provides more info, the legend label could just have the AUC value (or no legend)

glemaitre replied:
I would not change the default behaviour but I agree that maybe we would not need any legend.

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -129,25 +163,40 @@ def plot(
-            default_line_kwargs["label"] = f"AUC = {self.roc_auc:0.2f}"
-        elif name is not None:
-            default_line_kwargs["label"] = name
+        # TODO: Not sure about this, as ideally we would check params are correct
+        # first??
+        self.ax_, self.figure_, name_ = self._validate_plot_params(ax=ax, name=name)
```
Ideally we do parameter checking first, but I think this is okay...?

lucyleeow committed:
process attrs in plot

lucyleeow commented:
Thanks @glemaitre, changes made. We process in `plot` now and have both e.g., `self.tpr` and `self.tpr_`

lucyleeow committed:
add ref

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -14,25 +20,37 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
     stored as attributes.
 
-    Read more in the :ref:`User Guide <visualizations>`.
+    For more about the ROC metric, see :ref:`roc_metrics`.
+    For more about scikit-learn visualization classes, see :ref:`visualizations`.
```
This adds reference the user guide section on roc metrics and visualization classes.

I initially wanted to follow the 'Read more in our user guide' line that we use everywhere else, but since we are reference 2 places in the user guide, i thought it would be okay to change for this case but I am not 100%. Happy to amend.

glemaitre reviewed (commented):
OK here is a bunch of comments.

glemaitre commented:
But overall it looks quite good I think. @jeremiedbb do you have additional remarks?

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -14,25 +20,37 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
+        and `tpr`. If `None`, no area under ROC curve score is shown. If `name`
+        is also `None` no legend is added.
+
+    name : str or list of str, default=None
```
I have forgotten where we were up to in our discussions with this. Are we happy with the term 'name' ? I think it is reasonable.

And yes it is a breaking change so we should deprecate `estimator_name` and add `name` right? And I can add a function that errors if both params are given and warns if the deprecated one is.

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -129,25 +164,48 @@ def plot(
 
             .. versionadded:: 1.6
 
+        fold_line_kwargs : dict or list of dict, default=None
```
Sorry I do not follow this.

In the example above, do you mean `fold_line_kwargs` instead of `chance_level_kw` ?

If you mean `chance_level_kw`, there is only one per plot, so you only need to pass this once?

If you mean `fold_line_kwargs`, you can just pass one set of kwargs and `fold_line_kwargs` will apply it to all curves. Unless you mean we want to pass some default kwargs that apply to all curves, and then also some specific ones (e.g., colour) that should be applied to each curve?

lucyleeow replied:
note from our discussion - the idea is to be able to set some default line kwargs to apply to all curves (incl chance) and then also be able to set per curve kwargs that over-write default

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -129,25 +164,48 @@ def plot(
-            default_line_kwargs["label"] = f"AUC = {self.roc_auc:0.2f}"
-        elif name is not None:
-            default_line_kwargs["label"] = name
+        self.fpr_ = _convert_to_list_leaving_none(self.fpr)
+        self.tpr_ = _convert_to_list_leaving_none(self.tpr)
+        self.roc_auc_ = _convert_to_list_leaving_none(self.roc_auc)
```
The problem is that `_BinaryClassifierCurveDisplayMixin` is used for many display classes not just ROC curve.

We could have a `**kwarg` at the end, where all these parameters are validated with `_convert_to_list_leaving_none` BUT we have to set a e.g., `self.tpr_` and I don't know how you would do this...?

lucyleeow replied:
Clarification from discussion - Idea is to have a specific `_validate_plot_params` for individual display classes that calls `super._validate_plot_params` and then performs class specific actions

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -129,25 +164,48 @@ def plot(
+        self.ax_, self.figure_, self.name_ = self._validate_plot_params(
+            ax=ax, name=name
+        )
+        self.name_ = _convert_to_list_leaving_none(self.name_)
```
I think this is easier but will require a 'not so pretty':
```
if self.__class__.__name__ in (<classes requiring name_>):
    self.name_ = _convert_to_list_leaving_none()
```

lucyleeow replied:
Done

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -440,15 +504,186 @@ def from_predictions(
+                response_method=response_method,
+                pos_label=pos_label,
+            )[0]
+            # Should we use `_validate_from_predictions_params` here?
```
I put this comment in the wrong place, it should be above `_get_response_values_binary`.
Also I meant `_validate_and_get_response_values`.

`_validate_and_get_response_values` calls:

1. check_matplotlib_support
2. updates name to estimator name if name is None
3. `_get_response_values_binary`

1 only needs to be called once, but I guess it's not too much overhead if we call it a few times?
2 is not relevant here because we're using cv results, so estimator name is the same for all curves and not what we want to use to name the curves
3 this is what we really need

I think `_get_response_values_binary` validates `pos_label`, so we are not missing that. I think I only put this comment here for the `check_matplotlib_support` and so it is consistent with the other methods (e.g., if we change something we only need to change it one place). Maybe for this reason we should use `_get_response_values_binary` but ignore the `name` it returns...? Could be swayed either way.

lucyleeow replied:
Added a `_validate_from_cv_results_params` method that does all param checking and `check_matplotlib_support`

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -440,15 +504,186 @@ def from_predictions(
+                y_true,
+                y_pred,
+                pos_label=pos_label,
+                sample_weight=sample_weight,
```
`_validate_from_predictions_params` only checks that `sample_weight` is the same length as `y_true` and `y_pred`, 
But you are right, this is a bug and I should be using `_safe_indexing`

I think the problem is that `_validation_from_predictions_param` and `_validate_and_get_response_values` are very specific for the `from_estimator` and `from_prediction` methods.

Maybe it is a good idea to add a `_validate_cv_result_params` method?

lucyleeow replied:
Added a `_validate_cv_result_params`

lucyleeow committed:
Merge branch 'main' into cv_results3

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -129,25 +164,48 @@ def plot(
+            ax=ax, name=name
+        )
+        self.name_ = _convert_to_list_leaving_none(self.name_)
+        _check_param_lengths(
```
done

lucyleeow committed:
review and add tests

lucyleeow commented on sklearn/utils/_plotting.py:
```diff
@@ -24,9 +26,74 @@ def _validate_plot_params(self, *, ax=None, name=None):
+        # if name is None:
+        #     name = getattr(self, "estimator_name", None)
+        # if name is None:
+        #     name = getattr(self, "names", None)
```
This is nice, thank you

lucyleeow commented on sklearn/metrics/_plot/tests/test_common_curve_display.py:
```diff
@@ -176,9 +174,9 @@ def test_display_curve_estimator_name_multiple_calls(
-    "Display", [DetCurveDisplay, PrecisionRecallDisplay, RocCurveDisplay]
-)
+# Add separate test for displays that have converted to names?,
+# add note to remove this one in 1.9
```
Yes I think you are right.

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -129,25 +164,48 @@ def plot(
             .. versionadded:: 1.6
 
+        fold_line_kwargs : dict or list of dict, default=None
+            Dictionary with keywords passed to the matplotlib's `plot` function
+            to draw the individual ROC curves. If a list is provided, the
+            parameters are applied to the ROC curves sequentially. If a single
+            dictionary is provided, the same parameters are applied to all ROC
+            curves. Ignored for single curve plots - pass as `**kwargs` for
+            single curve plots.
+
+            .. versionadded:: 1.7
+
         **kwargs : dict
-            Keyword arguments to be passed to matplotlib's `plot`.
+            For a single curve plots only, keyword arguments to be passed to
+            matplotlib's `plot`. Ignored for multi-curve plots - use `fold_line_kwargs`
+            for multi-curve plots.
```
I completely agree, and I find it also annoying because I can't put the parameters in a better order (as I wanted to keep both the line kwarg params together) BUT I could not think of a way to allow this while being backwards compatible (allow users to also pass kwargs that get passed to mpl for case of single curve). I may be missing something here though?

jeremiedbb replied:
I opened https://github.com/lucyleeow/scikit-learn/pull/2 to your branch to deprecate `**kwargs` and add `line_kwargs` instead of `fold_line_kwargs` which applies to all kinds of curve.

I doesn't seem problematic from a backward compat point of view, but you might have use cases in mind that I did not thought about.

However if it appears to do the job, I think we could extract this deprecation from this PR and make it as a separate PR to ease the review here.

lucyleeow replied:
> I doesn't seem problematic from a backward compat point of view

I would personally prefer this (but I am worried I am not being nice enough to users).
No other cases, simply just that we need to deprecate. It would also be different to all our other display classes that allow kwargs - I imagine we would add `from_cv_results` to some other classes and also deprecate kwargs, but not all. The result would be that we have some 'multi curve' display classes that don't take kwargs and non-multi display classes that do take kwargs ?

I am okay with this, but just wanted to point this out in case someone else opposes.

lucyleeow replied:
Thinking again, it would be confusing to users who only wish to plot one curve, that some of our display objects take kwargs and others don't, for reasons that they may not grasp.

Is it reasonable to make all display objects take `line_kwargs` instead of `**kwargs` ?

jeremiedbb replied:
I'd rather have ``line_kwargs`` in all displays. Having displays with ``line_kwargs`` and displays with ``**kwargs`` is not very user friendly indeed. What do you think @glemaitre ?

lucyleeow replied:
Small detail but it would be more consistent if we called the plot lines either "curve" (if we decide on `curve_name`) or "line" (e.g., "line_kwargs") but not both...WDYT ?

lucyleeow replied:
Thinking about it more, I am not sure about the format `<name>_line_kwargs` , and I think I would prefer just `line_kwargs` because:

* not all display objects have an short name / recognisable abbreviation / have 'curve' in the name like `roc` - e.g., do we use `calibration_line_kwargs` or `calibration_curve_line_kwargs` / `db_line_kwargs` or `decision_boundary_line_kwargs` ?
* for some reason I can't put my finger on I think it would be nicer if this param name was the same across all display objects?

lucyleeow replied:
Talking with @glemaitre , we think `curve_kwargs` and `curve_name`

jeremiedbb replied:
`curve_kwargs` looks good to me

lucyleeow commented on sklearn/utils/_plotting.py:
```diff
@@ -24,9 +25,66 @@ def _validate_plot_params(self, *, ax=None, name=None):
+            fold_line_kws = [kwargs]
+        else:
+            if fold_line_kws is None:
+                fold_line_kws = [{"alpha": 0.5, "linestyle": "--"}] * n_curves
```
Yes setting reasonable default is difficult. I guess if cv is over ~5-10, different colours make things hard to follow. At the start of the PR we set all lines to be the same color but then we realised there would be no point labelling in the legend then -though I would be fine with this.  If the purpose of this plot is to help user visualise overall variation/some kind of 'variance', then single colour lines would be reasonable.

@glemaitre probably has more thoughts on this.

glemaitre replied:
We probably have two ways to go here:

- make each line distinguishable from another and we expect the legend for each line
- do not distinguish each line from another and have a single entry in the legend with the mean + std. dev. of the AUC

Here, an important point is that we want to have the best default. However, we should allow user to tweak the default.

Maybe we could go for solution 2) as default because it allows to always have a default that works whatever the parameter of the cross-validation.

jeremiedbb replied:
Another motivation for solution 2 is if one wants to compare 2 models. Having individual colors per curve is not helpful and even makes the plot messier. And if we have 10 curves per estimator, matplotlib even loops through the same colors so we can't distinguish the models any more. Look at this example
```py
import matplotlib.pyplot as plt
from sklearn.datasets import make_classification
from sklearn.metrics import RocCurveDisplay
from sklearn.model_selection import cross_validate
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression

X, y = make_classification(random_state=0, n_samples=1000, n_features=10, n_informative=3)

clf = SVC(random_state=0)
clf2 = LogisticRegression(random_state=0)

cv_results = cross_validate(clf, X, y, cv=10, return_estimator=True, return_indices=True)
cv_results2 = cross_validate(clf2, X, y, cv=10, return_estimator=True, return_indices=True)

fig, ax = fig, ax = plt.subplots()
RocCurveDisplay.from_cv_results(cv_results, X, y, ax=ax)
RocCurveDisplay.from_cv_results(cv_results2, X, y, ax=ax)
plt.show()
```

Also, if some day we decide to implement and display and averaged roc curve, having individual colors per curve would again not be very helpful. And what color would be the average ? probably black but would not look so great imo.

A third point is that as soon as there are more than 10 curves, the legend takes too much space and hides the curves, unless making a very large figure.

> do not distinguish each line from another and have a single entry in the legend with the mean + std. dev. of the AUC

I like the idea of a single entry with mean + std. For all the reasons listed above, I think it's probably a better default.

lucyleeow replied:
I am happy with this solution.

Naive question but is average a reasonable statistic to use for summarising roc auc ? Just wondering if we will run into any statistical poor practice by doing this.

glemaitre replied:
A mean ROC curve would be problematic. The mean + std. dev. of the AUC should be less problematic. I have to check the following publication to be sure that we are not having an issue: https://openreview.net/forum?id=FByH3qL87G

lucyleeow replied:
I had a quick re-read of the paper and they do not make any points against averaging ROC AUC, except that it alone does not tell you the full picture.

I will amend this such that we only add one item in the legend e.g. 'Average AUC: ' , which will allow you to compare between CVs with different hyperparameters or different estimators.  I won't add the name of the estimator, just because the user may be comparing between different hyperparameters of the same estimator.

Happy to change this though.

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -440,15 +504,186 @@ def from_predictions(
+                "and return_indices=True to the function cross_validate."
+            )
+
+        train_size, test_size = (
```
Yeah, I was hoping to add tests at a stage when we settled more or less on large changes (like the change to plural and then back :sweat_smile: ). I think we are past that stage and i have added tests now, which did make me re-think some validation and fix some things.

lucyleeow committed:
review and fix tests

lucyleeow commented:
@glemaitre @jeremiedbb I think I've addressed all comments and I've added tests which prompted me to improve the validation functions. I think it is ready for another look!

lucyleeow committed:
comment

glemaitre commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -14,25 +20,38 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
+        is also `None` no legend is added.
+
+    name : str or list of str, default=None
+        (Do we prefer curve_name) ?
```
`name` is a bit generic while `curve_name` is slightly more to the point. @jeremiedbb what do you think.

jeremiedbb replied:
``curve_name`` works for me. Maybe we could extract this deprecation from the PR and make it a separate PR to ease the review ?

glemaitre commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -440,15 +516,187 @@ def from_predictions(
 
         return viz.plot(
             ax=ax,
+            # Should we provide `name` to both `cls` and `plot` or just `cls`?
```
Passing to the constructor should be enough.

lucyleeow replied:
I've amended this for `from_predictions` as well (only pass to `cls`), to be consistent, but happy to revert.

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -177,3 +319,80 @@ def _despine(ax):
+    version = parse_version(version)
+    # Not sure if I should hard code this because this wouldn't work if we release
+    # a new major version ?
+    version_remove = f"{version.major}.{version.minor + 2}"
```
I think it is good enough for the moment. If we are going to release a major one then we would have to do many changes everywhere, so it will be part of it.

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -177,3 +319,80 @@ def _despine(ax):
+        )
+
+
+# Potentially useful for non binary displays `LearningCurveDisplay` and
+# `ValidationCurveDisplay`, so not placed under `_BinaryClassifierCurveDisplayMixin`
```
Let's move it in the mixin for the moment and move it away when we see that it is useful outside.

You can add a `# TODO` to kind of remember about it.

lucyleeow committed:
merge main

lucyleeow committed:
review and add tests

lucyleeow commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +75,201 @@ def _validate_from_predictions_params(
+    # TODO: if useful for non binary displays (e.g.,`LearningCurveDisplay`,
+    # `ValidationCurveDisplay`) amend to function
+    @classmethod
+    def _validate_line_kwargs(
```
I am wondering if it might be nice to test these class methods in isolation, to avoid repeating the same/similar tests in each child class - but may be it is better to test it in the child classes to make sure it works all together...?

cc @glemaitre

glemaitre replied:
We could test them. I think they should not be a class method but rather a static method because I don't see any use of cls

lucyleeow committed:
add aggregate legend

lucyleeow commented:
I have made the default behaviour to add a single legend entry of format "AUC = mean +/- std". There were a lot more API decisions than anticipated (and it was way more complicated). Summary of decisions, which can be changed:

* Added a explicit `show_aggregate_score` parameter to `from_cv_results` - I could have made it implicit in that if `name` or `fold_line_kwargs` was a single item, it meant that a single legend entry with mean/std was to be added, but I think being explicit is better
* for `from_cv_results`, if `show_aggregate_score` is True:
   * `name` (updated from `fold_names` as this is no longer relevant) - has to be None or a string or list of length 1
   * `fold_line_kwargs` - has to be a dict (should we also allow a list of 1 dict too? it's extra complexity for us and was never part of original API)
* for `from_cv_results`, if `show_aggregate_score` is False:
   * `name` has to be None or list of length `n_curves`
   * no `fold_line_kwargs` restrictions here
* for `plot`, if `roc_auc_aggregate` attr is provided:
   * `name` (updated from `fold_names` as this is no longer relevant) - has to be None or a string or list of length 1
   * `fold_line_kwargs` - has to be a dict (should we also allow a list of 1 dict too??)

I am wondering if we should allow `plot` to be more flexible in regards to `fold_line_kwargs`:
* pro: if the user is using `plot` they may be more experienced and we should allow them to do what they want
* con: with a single legend entry, it would be confusing if the curves all looked different as we 'label' the first curve only, to achieve a single legend entry. 

I've amended most tests so they pass, and commented 2 tests out that require more thought. I anticipate API changes/discussion so have not updated these 2 tests nor added new tests. Will do this once we're more settled on API.

cc @glemaitre @jeremiedbb

lucyleeow committed:
amend condition

jeremiedbb commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -440,15 +561,203 @@ def from_predictions(
+            sequentially. If a single dictionary is provided, the same
+            parameters are applied to all ROC curves.
+
+        show_aggregate_score : bool, default=True
+            Whether to show the ROC AUC mean and standard deviation of curves from
+            all folds as a single legend entry. If `True`, `name` should be a single
+            string and `fold_line_kwargs` should be a single dictionary, to prevent
+            confusion in the legend. If `False`, `name` should be None or a list the
+            same length as the number of cross-validation folds.
```
I don't think that we need to introduce this parameter. We can just automatically aggregate if ``name`` and ``line_kwargs`` are not lists and not aggregate otherwise.

lucyleeow replied:
I understand your concern but was I worried it would be confusing for the user (and explain in the docstring). Here is a summary of what the behaviour would be for all combinations:

* `name=None`, `line_kwargs=None` - aggregate (default behaviour)
* `name=None`, `line_kwargs=dict` - aggregate (dict applies to all curves)
* `name=None`, `line_kwargs=list[dict]` - don't aggregate
* `name=str`, `line_kwargs=None` - aggregate
* `name=list`, `line_kwargs=None` - don't aggregate
* `name=str`, `line_kwargs=dict` - aggregate
* `name=list`, `line_kwargs=list` - don't aggregate
Possibly (?) conflicting combinations:
* `name=list`, `line_kwargs=dict` - error ? (does not make sense to label individual curves that have the same appearance) 
* `name=str`, `line_kwargs=list[dict]` - don't aggregate and repeat names for each line (?)

Essentially, if one of `name` or `line_kwargs` is a list, don't aggregate.

It's not as bad as I originally thought, now that I've written it down. I'm happy to change to this behaviour if you wish?

jeremiedbb replied:
To me ``show_aggregate_score`` is completely redundant and is fully determined by the types of ``name`` and ``line_kwargs``, so I find it more confusing to add it.

lucyleeow replied:
I can see your point. Do you agree with the proposal above?

jeremiedbb replied:
Yep

> name=str, line_kwargs=list[dict] - don't aggregate and repeat names for each line (?) (note that the AUC will be different for each line so the label will be different) (maybe error here as well..?)

I think it's okay to allow it. No need to raise

lucyleeow replied:
Honestly this is not easy explain in the docstring.

I think we should make it possible for user to not add a legend - it makes sense that this happens when `name=None`. I can change this though.

lucyleeow replied:
* `name=None`, `line_kwargs=None` - no legend
* `name=None`, `line_kwargs=dict` - no legend
* `name=None`, `line_kwargs=list[dict]` - no legend
* `name=str`, `line_kwargs=None` - aggregate
* `name=list`, `line_kwargs=None` - don't aggregate
* `name=str`, `line_kwargs=dict` - aggregate
* `name=list`, `line_kwargs=list` - don't aggregate

Possibly (?) conflicting combinations:
* `name=list`, `line_kwargs=dict` - error ? (does not make sense to label individual curves that have the same appearance) 
* `name=str`, `line_kwargs=list[dict]` - don't aggregate and repeat names for each line (?) (note that the AUC will be different for each line so the label will be different) (maybe error here as well..?)

lucyleeow replied:
This is the `name` docstring ....

```
            Name for labeling legend entries.

            - str :

              - if `fold_line_kwargs` is None or a dictionary, label all curves using
                a single legend entry and the mean ROC AUC score
              - if `fold_line_kwargs` is a list of dictionaries, label individual
                curves with the same name and its individual ROC AUC score.

            - list of str : label each individual curve, with it's ROC AUC score.
              List should be the same length as the number of cross-validation folds.
              Cannot be used in conjunction with `fold_line_kwargs` being a single
              dictionary or None, to avoid labeling individual curves that have the
              same appearance.
            - `None` :

                - if `fold_line_kwargs` is a list of dictionaries, each individual
                  curve is present in the legend with it's ROC AUC score but no name.
                - if `fold_line_kwargs` is a single dictionary or None, a single
                  legend entry with the mean ROC AUC score but no name is shown.
```

happy to take suggestions. I could be more succinent and just specify the number of legend entries depends on `fold_line_kwargs` but then you sort of have to explain default `None` and name: str and `fold_line_kwargs`: List combination , behaviour...

lucyleeow replied:
Okay I've hit a snag.
`from_cv_results` has default line kwargs. Currently we incorporate these default line kwargs into `line_kwargs`, such that we always pass a list of length `n_curve` to `plot`, meaning it is impossible to tell whether we should aggregate or not, in `plot`.

Here are some options:

* Add a (private) `default_line_kwargs` to `plot` such that the incorporation is done in `plot` - it does require adding a additional parameter
* Determine if we aggregate in `from_cv_results` and pass this info to `plot` or the class via an additional parameter/private attribute
* something else ??

cc @jeremiedbb @glemaitre

jeremiedbb replied:
In https://github.com/lucyleeow/scikit-learn/pull/2 I moved the default line kwargs out of `from_cv_results` and included them in plot instead. This way the line kwargs that `plot` receives allows it to determine whether it should aggregate or not

jeremiedbb commented on doc/whats_new/upcoming_changes/sklearn.metrics/30399.enhancement.rst:
```diff
@@ -0,0 +1,4 @@
+- Add class method `from_cv_results` to :class:`metrics.RocCurveDisplay`, which allows
+  easy plotting of multiple ROC curves from :func:`model_selection.cross_validate`
+  results.
+  By :user:`Lucy Liu <lucyleeow>`
```
To me it qualifies as a feature

jeremiedbb commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -14,25 +21,53 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
+        If plotting multiple curves, list should be of same length as
+        and `tpr`.
+
+    tpr : ndarray list of ndarrays
```
```suggestion
    tpr : ndarray or list of ndarrays
```

jeremiedbb commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -14,25 +21,53 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
-        Name of estimator. If None, the estimator name is not shown.
+    fpr : ndarray or list of ndarrays
+        False positive rates. Each ndarray should contain values for a single curve.
+        If plotting multiple curves, list should be of same length as
+        and `tpr`.
```
```suggestion
        If plotting multiple curves, list should be of same length as `tpr`.
```

jeremiedbb commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -14,25 +21,53 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
+
+    tpr : ndarray list of ndarrays
+        True positive rates. Each ndarray should contain values for a single curve.
+        If plotting multiple curves, list should be of same length as
+        and `fpr`.
```
```suggestion
        If plotting multiple curves, list should be of same length as `fpr`.
```

jeremiedbb commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -14,25 +21,53 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
+        `roc_auc_aggregate` for alternative.
+        If `name` and `roc_auc_aggregate` are also `None` no legend is added.
+
+    roc_auc_aggregate : tuple(float), default=None
+        ROC AUC mean and standard deviation. An alternative to `roc_auc` when
+        plotting multiple curves and a single legend entry showing ROC AUC mean and
+        standard deviation for all curves is desired.
+        If `True`, `name` cannot be a list of length >1.
+
```
I don't think that we need to introduce this parameter. Similar to `show_aggregate_score`, I think it's redundant and brings to much coupling between parameters.

Instead, I think that deciding on showing an aggregate or not should be done automatically in `plot`, depending on whether `curve_name` or `roc_line_kwargs` are lists or not.

lucyleeow committed:
Merge branch 'main' into cv_results3

lucyleeow committed:
review

lucyleeow committed:
wip remove agg params

lucyleeow committed:
iter remove aggre param

lucyleeow committed:
remove agg, tests pass

lucyleeow commented:
Amended code to remove the "aggregate" parameters.

* `line_kwargs` now determines whether to label individual curves or single aggregate label - `list` means to label individual curves.
   * `line_kwargs` values that result in lines looking the same cannot be used with list `name`, as there is no point individualy labelling each curve, when all curves look the same
* pass default line kwargs (from `from_cv_results`) through to `plot` as a single dict - this enables us to still be able to infer whether to aggregate label, without requiring an additional parameter
   * not sure if https://github.com/scikit-learn/scikit-learn/pull/30399#discussion_r2026623754 works as the default line kwargs (`{"alpha": 0.5, "linestyle": "--"}`) don't seem to be passed?

The tests pass (bar the previously commented out ones, which require more thought/work) and the code is at a stage where it is ready to be checked to see if you are happy with the overall implementation/API @jeremiedbb @glemaitre :pray: 

The aggregate and individual label has introduced a lot of fiddly conditions/code complexity, which I will think more about and try to clean up, but I would prefer to do this after we decide that we are happy with the implementation.
(Also need to amend the commented out tests.)

lucyleeow commented:
(Also very happy to deprecate **kwargs and amend names to `curve_name` and `curve_kwargs` in this PR, but just wanted to settle on API/implementation first)

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +74,242 @@ def _validate_from_predictions_params(
+    # TODO: if useful for non binary displays (e.g.,`LearningCurveDisplay`,
+    # `ValidationCurveDisplay`) amend to function
+    @classmethod
+    def _validate_line_kwargs(
```
Should the name reflect that it is especially for fold lines

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +74,242 @@ def _validate_from_predictions_params(
+        return label
+
+    @classmethod
+    def _get_line_kwargs(
```
Usually we use validate instead of get as a name

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +74,242 @@ def _validate_from_predictions_params(
+        n_curves : int
+            Number of curves.
+
+        name : list[str] or None
```
Nitpick for later: we need to follow out own standard for the docstring.

jeremiedbb commented:
> Amended code to remove the "aggregate" parameters.

The diff still shows them though. Maybe you forgot to push your last commits ?

> line_kwargs now determines whether to label individual curves or single aggregate label - list means to label individual curves.
    line_kwargs values that result in lines looking the same cannot be used with list name, as there is no point individualy labelling each curve, when all curves look the same

+1

> pass default line kwargs (from from_cv_results) through to plot as a single dict - this enables us to still be able to infer whether to aggregate label, without requiring an additional parameter
not sure if https://github.com/scikit-learn/scikit-learn/pull/30399#discussion_r2026623754 works as the default line kwargs ({"alpha": 0.5, "linestyle": "--"}) don't seem to be passed?

I think we can define the default style directly in ``plot``, or even in the validate method and not have to pass it around. That's what I do in https://github.com/lucyleeow/scikit-learn/pull/2.

>  the code is at a stage where it is ready to be checked to see if you are happy with the overall implementation/API

Hard to tell since the diff doesn't yet all the API points that you discussed but I think I'm happy with what comes out of the different discussions, that is:

- By default, all curves have the same style: blue plain line. A single legend shows mean+std AUC.

- don't introduce "\*aggregate\*" parameters neither in ``from_cv_results`` nor in ``plot`` and instead automatically infer the style to adopt based on ``curve_kwargs`` and ``curve_name``.

- deprecate **kwargs for single curve in favor of ``curve_kwargs``. Same for ``curve_name``. This can be done in 1 or 2 separate PRs to decompose into more simple changes, or not if it makes thing more complicated.

> The aggregate and individual label has introduced a lot of fiddly conditions/code complexity, which I will think more about and try to clean up

I also think that the added complexity for the validations is a lot higher than it should be, but I'm confident that we'll be able to reduce it back to a reasonable state when all these API changes are implemented.

lucyleeow committed:
rm old param

lucyleeow commented:
:facepalm: I don't know how I managed that but yes you are correct - I forgot to push the last commits. So sorry. Would still really appreciate a review :pray: 

> I think we can define the default style directly in plot, or even in the validate method and not have to pass it around. That's what I do in https://github.com/lucyleeow/scikit-learn/pull/2.

It's not passed around as a separate parameter, it's passed to `plot` inside `from_cv_results` but in a way that we can still infer aggregate status (i.e. as a dict and not a list, which means 'aggregate')

I don't see the actual default parameters being passed in that PR..?

Also I am afraid the new changes will create lots of merge conflicts in that PR. I am happy to just incorporate the deprecation and name changes in this PR, because fixing this merge conflicts may be unpleasant.

jeremiedbb commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +80,198 @@ def _validate_from_predictions_params(
+                )
+
+        # Add `default_curve_kwargs` to `fold_line_kwargs`
+        default_curve_kwargs = {"alpha": 0.5, "linestyle": "--"}
```
```suggestion
        default_curve_kwargs = {"alpha": 0.5, "linestyle": "-", "color": "blue"}
```

jeremiedbb commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +80,198 @@ def _validate_from_predictions_params(
+        default_curve_kwargs = {"alpha": 0.5, "linestyle": "--"}
+        if fold_line_kwargs is None:
+            fold_line_kwargs_ = default_curve_kwargs
+        elif isinstance(fold_line_kwargs, Mapping):
+            fold_line_kwargs_ = _validate_style_kwargs(
+                default_curve_kwargs, fold_line_kwargs
```
With this approach, `from_cv_results` calls `validate_from_cv_results_params` which calls `validate_style_kwargs`. Then it calls `plot` which calls `get_line_kwargs` which again calls `validate_style_kwargs`. There's duplicated validation.

That's why I suggested to have all validation of the curve kwargs and incorporation of defaults happening in plot. `from_predictions` doesn't validate the kwargs and directly passes them to `plot`. It would be cleaner and simpler to have a single centralized place to validate the curve kwargs for all methods.

lucyleeow replied:
Thinking again, I think it makes sense to have these default kwargs always when a user is plotting multiple curves, so will amend

jeremiedbb commented:
> Also I am afraid the new changes will create lots of merge conflicts in that PR. I am happy to just incorporate the deprecation and name changes in this PR, because fixing these merge conflicts may be unpleasant.

Alright

glemaitre reviewed (commented):
I'm fine with the general API. I only would think about some naming that I could find easier to follow and maybe a couple of type that I think that could make it also easier to understand the code. But in terms of public exposed API, it looks good.

Thanks @lucyleeow for the patience.

lucyleeow committed:
guillaume review

lucyleeow committed:
validate only in plot

lucyleeow committed:
name change to curve_kwargs

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -76,19 +120,52 @@ class RocCurveDisplay(_BinaryClassifierCurveDisplayMixin):
+    def _validate_plot_params(self, *, ax, name, fold_line_kwargs):
+        self.ax_, self.figure_, name_ = super()._validate_plot_params(ax=ax, name=name)
+
+        self.fpr_ = _convert_to_list_leaving_none(self.fpr)
+        self.tpr_ = _convert_to_list_leaving_none(self.tpr)
+        self.roc_auc_ = _convert_to_list_leaving_none(self.roc_auc)
+        self.name_ = _convert_to_list_leaving_none(name_)
```
No particular reason now that I think about it. Amended.

lucyleeow committed:
pass curve_kwarg properly and order in docstring

lucyleeow committed:
fix naming

lucyleeow committed:
add kwarg dep test

lucyleeow committed:
guillaume review

lucyleeow committed:
fix test

lucyleeow committed:
format

lucyleeow committed:
format

lucyleeow committed:
format

lucyleeow committed:
format

lucyleeow committed:
revert **kwargs

lucyleeow committed:
merge main

lucyleeow committed:
add tests codecov

lucyleeow committed:
add _BinaryClassifierCurveDisplayMixin tests

lucyleeow committed:
typo

lucyleeow commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +69,218 @@ def _validate_from_predictions_params(
+            # Adapt error message
+            raise ValueError(str(e).replace("y_true", "y"))
+
+        # n_curves = len(cv_results["estimator"])
+        # # NB: Both these also checked in `plot`, but thought it best to fail earlier.
+        # cls._validate_multi_curve_kwargs(cls, curve_kwargs, name, n_curves)
+        # if isinstance(name, list) and len(name) not in (1, n_curves):
+        #     raise ValueError(
+        #         f"`name` must be None, a list of length {n_curves} or a single "
+        #         f"string. Got list of length: {len(name)}."
+        #     )
```
I felt that we should 'fail earlier' for this check and also the `**kwarg` deprecation (which is done in `_validate_curve_kwargs`, which not run until `plot`. In some cases we do a bit of compute before these 2 checks are run. BUT doing this checking earlier does result in redundant validation as everything is eventually passed to `plot`. 

WDYT @jeremiedbb @glemaitre

lucyleeow replied:
After discussing with @glemaitre , we decided that while it would be nice to fail earlier, it is okay here that the validation is delayed until `plot`.

Ultimately, it is due to how Displays are implemented. Potentially if we had implemented it such that the call to plot was separate, e.g.,
```python
display = RocCurveDisplay.from_estimator(kwargs)
display.plot(kwargs)
```

Then the plot kwargs could be validated at `plot`, but oh well!

lucyleeow committed:
change to feature

lucyleeow commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -134,22 +225,34 @@ def plot(
             Keyword arguments to be passed to matplotlib's `plot`.
 
+            .. deprecated:: 1.7
+                kwargs is deprecated and will be removed in 1.9. Pass matplotlib
```
Using "**kwargs" here instead of "kwargs" results in CI failure due to a docutils check, but I think this is fine.

lucyleeow reviewed (commented):
Resolved some nightmar-ish merge conflicts due to #29865, hopefully all okay.

I still want to add a unit test for the static method `_validate_curve_kwargs` but had a question first.

lucyleeow committed:
remove early param validation

lucyleeow committed:
Merge branch 'main' into cv_results3

lucyleeow committed:
lint

glemaitre commented:
Let me do another round of review.

lucyleeow committed:
Merge branch 'main' into cv_results3

glemaitre commented on sklearn/utils/tests/test_plotting.py:
```diff
@@ -2,13 +2,111 @@
+    label = _BinaryClassifierCurveDisplayMixin._get_legend_label(
+        curve_legend_metric, curve_name, legend_metric_name
+    )
+    print(label)
```
```suggestion
```

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +69,185 @@ def _validate_from_predictions_params(
+
+        curve_kwargs : dict or list of dict
+            Dictionary with keywords passed to the matplotlib's `plot` function
+            to draw the individual ROC curves. If a list is provided, the
```
```suggestion
            to draw the individual curves. If a list is provided, the
```

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +69,185 @@ def _validate_from_predictions_params(
+            Dictionary with keywords passed to the matplotlib's `plot` function
+            to draw the individual ROC curves. If a list is provided, the
+            parameters are applied to the ROC curves sequentially. If a single
+            dictionary is provided, the same parameters are applied to all ROC
```
```suggestion
            dictionary is provided, the same parameters are applied to all
```

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +69,185 @@ def _validate_from_predictions_params(
+        **kwargs : dict
+            Deprecated. Keyword arguments to be passed to matplotlib's `plot`.
+        """
+        # Deprecate **kwargs
```
Can you add a `# TODO(1.9): xxx` such that we remember to remove this part of the code later on.

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +69,185 @@ def _validate_from_predictions_params(
+            warnings.warn(
+                "`**kwargs` is deprecated and will be removed in 1.9. Pass all "
+                "matplotlib arguments to `curve_kwargs` as a dictionary instead.",
+                FutureWarning,
```
I don't see a unit test in `test_plotting.py`. We can add it also and have a `# TODO` there as well.

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +69,185 @@ def _validate_from_predictions_params(
+                legend_metric["mean"], name[0], legend_metric_name
+            )
+            # Add the "+/- std" to the end (in brackets if name provided)
+            if legend_metric["mean"] is not None:
```
I have to check the usage but is weird to check for `legend_metric["mean"]` while do something with `legend_metric["std"]`. Could we check directly the `"std"` key?

lucyleeow replied:
Ah yes, technically both 'mean' and 'std' should be `None` or a number, (it does not work if they are not both the same).

Happy to change to 'std', doesn't really matter, but I wouldn't want people to think that there could be a case where 'mean' is provided by 'std' isn't (this case does not make sense as the legend would have a single 'mean' value but it would not be clear it is a mean, as we use "+/- std" to relate the fact that this value is a mean)

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -63,6 +69,185 @@ def _validate_from_predictions_params(
+                    )
+                )
+
+        curve_kwargs_ = []
+        for fold_idx, label in enumerate(labels):
+            label_kwarg = {"label": label}
+            curve_kwargs_.append(
+                _validate_style_kwargs(label_kwarg, curve_kwargs[fold_idx])
+            )
```
I think that a list comprehension here would still be readable.
```suggestion
        curve_kwargs_ = [
            _validate_style_kwargs({"label": label}, curve_kwargs[fold_idx])
            for fold_idx, label in enumerate(labels)
        ]
```

glemaitre replied:
we could even directly return the list comprehension:

```python
        return [
            _validate_style_kwargs({"label": label}, curve_kwargs[fold_idx])
            for fold_idx, label in enumerate(labels)
        ]
```

glemaitre commented on sklearn/utils/_plotting.py:
```diff
@@ -177,3 +362,57 @@ def _despine(ax):
         ax.spines[s].set_bounds(0, 1)
+
+
+def _deprecate_estimator_name(old, new, version):
```
Could you make quick unit tests for the utilities below?

glemaitre commented on sklearn/metrics/_plot/roc_curve.py:
```diff
@@ -164,7 +267,13 @@ def plot(
-        (self.line_,) = self.ax_.plot(self.fpr, self.tpr, **line_kwargs)
+        self.line_ = []
+        for fpr, tpr, line_kw in zip(fpr, tpr, curve_kwargs):
+            self.line_.extend(self.ax_.plot(fpr, tpr, **line_kw))
```
`self.ax_.plot(...)` will return a list. So if I'm not mistaken, we should have `self.line_` to be a list of list of Line2D. It is not inline with the documentation. I'm wondering if we should not have something like:


```suggestion
            self.line_.extend(self.ax_.plot(fpr, tpr, **line_kw)[0])
```

glemaitre replied:
But I might miss something. I should check if we have a test checking the attribute.

glemaitre commented on sklearn/metrics/_plot/tests/test_roc_curve_display.py:
```diff
@@ -24,12 +27,32 @@ def data():
     return X, y
 
 
+# This data always (with and without `drop_intermediate`)
```
Interesting point. I think that we should indeed avoid. But let's revisit after the PR.

lucyleeow replied:
Forgot about this, glad I wrote the note. Let me create an issue so we don't forget.

glemaitre commented on sklearn/metrics/_plot/tests/test_roc_curve_display.py:
```diff
@@ -103,27 +126,504 @@ def test_roc_curve_display_plotting(
+        )
+
+
+# TODO : Remove in 1.9
```
I think that we use this pattern
```suggestion
# TODO(1.9): Remove in 1.9
```

glemaitre commented on sklearn/metrics/_plot/tests/test_roc_curve_display.py:
```diff
@@ -103,27 +126,504 @@ def test_roc_curve_display_plotting(
+        RocCurveDisplay(fpr=fpr, tpr=tpr, estimator_name="test")
+
+
+# TODO : Remove in 1.9
```
```suggestion
# TODO(1.9): Remove in 1.9
```

glemaitre commented on sklearn/metrics/_plot/tests/test_roc_curve_display.py:
```diff
@@ -103,27 +126,504 @@ def test_roc_curve_display_plotting(
+    _check_figure_axes_and_labels(display, pos_label)
+    aggregate_expected_labels = ["AUC = 1.00 +/- 0.00", "_child1", "_child2"]
+    for idx, line in enumerate(display.line_):
+        assert isinstance(line, mpl.lines.Line2D)
```
Ao I should be wrong in my previous comment because we have the proper test :)

glemaitre commented on sklearn/metrics/_plot/tests/test_roc_curve_display.py:
```diff
@@ -223,6 +696,62 @@ def test_roc_curve_chance_level_line(
+        {"lw": 1, "color": "blue", "ls": "-", "label": None},
+    ],
+)
+# To ensure both curve line kwargs and change line kwargs passed correctly
```
Put this comment in the docstring of the test maybe.

glemaitre reviewed (commented):
Only a couple of nitpicks. It looks great on my side and almost ready to be merged.

lucyleeow committed:
review

lucyleeow committed:
add multi test

lucyleeow commented:
I think I have addressed everything, thanks @glemaitre !

lucyleeow committed:
merge main

lucyleeow committed:
fix tests for new data

lucyleeow committed:
merge main

lucyleeow commented:
@jeremiedbb gentle ping on this, thanks!

jeremiedbb committed:
nitpicks

jeremiedbb reviewed (approved):
I pushed tiny nitpicks.

LGTM. Thanks !

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/lucyleeow/30399

lucyleeow commented:
Thanks @jeremiedbb !
